api/src/main/java/org/pmiops/workbench/utils/WorkspaceMapper.java

```diff
@@ -199,4 +203,25 @@ WorkspaceResource dbWorkspaceAndDbDatasetToWorkspaceResource(
   default String cdrVersionId(CdrVersion cdrVersion) {
     return String.valueOf(cdrVersion.getCdrVersionId());
   }
+
```

Comment:
I thought simplicity in mappers was one of your motivations you mentioned for this. Am I missing something here? The trade offs I see overall are:

+ Improves clarity of which fields are required during creation, and offers more breathing room to clearly document implications of a field on create
- Duplicates a subset of the fields in the workspace object, which will need to be kept synchronized by hand and documented twice

The alternative to this approach would be to improve the documentation on the workspace object, and add the read-only labels as needed. This is my preference still, since the workspace object is really not that unusual. It has many read-only fields, yes, but this is pretty standard in a REST API.

Reply:
The problems I was running into had to do with fields that were not only optional, but not read at all during `createWorkspce`, but that were thereafter available (read-only for the purposes of that API). I've not yet run down whether any of those read-only fields are writable for updateWorkspace or any other places taking that object.

I agree that the duplication in the Swagger is undesirable. The hope is that we can consolidate somewhat with OpenApi 3. YAML anchors may not work well here.

As for simplifying the mapper, there are mechanisms to get around this, but I don't really like them. The root cause always smelled like a bad object design, but I realize API payloads are special in that regard, and we don't always expect them to be 1:1 w/r/t entities.

Reply:
My preferences looking at all of this:
- readonly annotations seem very useful to apply to our API models
- I'd prefer the standard REST pattern for our create methods. I'm not sold on the benefits of this approach, and I'd want a strong argument to stray from the standard pattern i.e. what makes this endpoint so unique that it needn't conform to standard REST patterns (and the Google API style guide).

To play devil's advocate a bit, there is some precedent to use a different model here for symmetry with Firecloud's underlying CreateWorkspace API (WorkspaceIngest), though I dislike that model for the same reasons.

RE anchors: I buy we might be able to avoid some duplication in the yaml, but that's only really half (or less) of the concern. There's duplication in the IDL, but then there's duplication in the generated code (tests, multiple overlapping models in the UI, etc). For example, we currently reuse the same model for the create / clone / edit pages on the UI. These are highly similar UI components which share a lot of code. The test stubs all use a single Workspace model throughout.

Reply:
I agree `CreateWorkspace` wasn't much of a win. The blast radius is huge and I think you're right that smart APIs and dumb models is the way to go.

I wouldn't want to use the existing Firecloud API pattern, as (1) it's probably going to change and (2) it's an implementation detail to our API client(s).

A bit about how I wound up here:

The test stubs were being sloppy and passing in post-created objects to the createWorkspace method, making it maddening to debug the tests. I think I still want separate stub methods for those, though many of the places in `WorkspacesControllerTest` that se `createWorkspace()` don't actually need to, and would be fine with a `static final Workspace`.

The reason I dug into the tests was to debug changes to `WorkspaceMapper`, which were confusing me for the same reason: so many fields are nullable vs read-only.

There's yet another branch where I have `readOnly` and `required` fields marked on the `Workspace` model. It sounds like that's the preferred path, all things considered. From an OOP standpoint I'm disgruntled, but I'll allow that it's hard to have cross-language patterns for reuse or inheritance.

So, closing this PR. May cannibalize some of the test fixes and any documentation additions. Will add `required` and `readOnly` to Workspace parameters and fixup front end code.

Reply:
> I thought simplicity in mappers was one of your motivations you mentioned for this.
The mapper changes are on another branch, and not yet updated for `CreateWorkspaceRequest`.